### ISSUE TITLE
Use a faster base64 implementation

### ISSUE THREAD

**Issue by ronaldtse** [riboseinc/rnp#16]
A very helpful base64 implementation [comparison](https://github.com/gaspardpetit/base64/).

**Comment by dewyatt** [comment]
I'd like to "upgrade" this to a bug since I found we apparently have multiple base64 implementations in use and at least one of them is broken.
The `frombase64` implementation in ssh2pgp.c appears to be subtly broken (and causes SSH key loading to fail for me).

**Comment by ni4** [comment]
Agree on that. Current base64 implementation is awfully slow, writing characters 1 by 1. You can assign this to me if nobody else want to dig into.

**Comment by ronaldtse** [comment]
Thanks @ni4 and @dewyatt !

**Comment by ronaldtse** [comment]
@riboseinc/rnp Anyone interested in handling this?

**Comment by flowher** [comment]
Yup. Me

**Comment by ni4** [comment]
@flowher you may go ahead, since 27 days left and I still didn't get to this issue, was occupied by other tasks.

**Comment by ronaldtse** [comment]
Thanks @flowher @ni4 👍

**Comment by ni4** [comment]
@flowher @ronaldtse  I implemented base64 encoding/decoding streams as part of PR#463 . So after mergin it this issue can be closed.

**Comment by flowher** [comment]
Any results?

**Comment by ni4** [comment]
@flowher Not yet, since only password encryption is now implemented (which wasn't implemented before). But code should be definitely faster.

**Comment by flowher** [comment]
Let's close it after we have some comparisons.

**Comment by dewyatt** [comment]
We may want to keep this issue open regardless, until we have removed the other base64 implementations (maybe).

**Comment by ni4** [comment]
@dewyatt @flowher Ok, agreed.

**Comment by flowher** [comment]
Another idea is to reuse botan's implementation of Base64

**Comment by flowher** [comment]
New base64 and CRC24 implementations seem to improve a lot the ASCII-armor encoding. I think we can close this record 

New implementation (28e358903064a4107c91b2c6f162d8a391074e28):
```
rnp --dearmor /tmp/data.arm --output /tmp/out  2.91s user 0.46s system 99% cpu 3.380 total
rnp --dearmor /tmp/data.arm --output /tmp/out  2.99s user 0.49s system 66% cpu 5.232 total
```
Old implementation (115f78e400213f1924802c600535b31ba89c71c0):
```
rnp --dearmor /tmp/data.arm --output /tmp/out  12.97s user 0.44s system 99% cpu 13.486 total
rnp --dearmor /tmp/data.arm --output /tmp/out  13.20s user 0.37s system 99% cpu 13.590 total
```

**Comment by ni4** [comment]
> We may want to keep this issue open regardless, until we have removed the other base64 implementations (maybe).

We still use old base64 implementation in key/keyring loading and saving. So I reopen this issue if nobody mind. I planned to fix that after finishing with streams.

**Comment by ronaldtse** [comment]
Sounds good indeed! Faster is almost always better...

**Comment by flowher** [comment]
@ni4: I'll do it then

**Comment by ni4** [comment]
@flowher Ok, thanks

**Comment by ni4** [comment]
Finally resolved in PR #684 - now single armoring/dearmoring code is used across the library. So closing this.
